### PR TITLE
Upgrade Calcite to 1.27

### PR DESCRIPTION
**Pull request by gierlachg**
Upgraded Calcite to latest (1.27).

Implemented support for `SARG` type/literal.
Reworked table functions - Calcite requires table functions to provide `SqlOperandMetadata`.
Increased maximum decimal precision to `76` - Calcite started validating decimals (see `SqlTypeUtil#isValidDecimalValue`). 

Had to patch `org.apache.calcite.linq4j.tree.ConstantExpression` - see [CALCITE-4532](https://issues.apache.org/jira/browse/CALCITE-4532) & https://github.com/apache/calcite/pull/2530.

### CODE REVIEW

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/misc/CaseOperatorIntegrationTest.java [hunk 1]
```diff
@@ -234,10 +234,10 @@ public void numericWithString() {
         checkFailure0(sql, SqlErrorCode.PARSING, "Cannot infer return type for CASE among [BIGINT, VARCHAR]");
 
         putBiValue(BigDecimal.ONE, "str", ExpressionTypes.BIG_DECIMAL, ExpressionTypes.STRING);
-        checkFailure0(sql, SqlErrorCode.PARSING, "Cannot infer return type for CASE among [DECIMAL(38, 38), VARCHAR]");
+        checkFailure0(sql, SqlErrorCode.PARSING, "Cannot infer return type for CASE among [DECIMAL(76, 38), VARCHAR]");
```
[the remark this answers is not in the document]
Comment (gierlachg) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/misc/CaseOperatorIntegrationTest.java:237]: > Increased maximum decimal precision to 76 - Calcite started validating decimals (see SqlTypeUtil#isValidDecimalValue).

**Review by viliam-durina — APPROVED** [review]

File: hazelcast-sql/src/main/codegen/config.fmpp [hunk 2]
```diff
@@ -470,4 +487,4 @@ data: {
 
 freemarkerLinks: {
   includes: includes/
-}
+}
```
Comment (hakanmemisoglu) [inline hazelcast-sql/src/main/codegen/config.fmpp:490]:
```suggestion
}

```

**Review by hakanmemisoglu — APPROVED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/operators/typeinference/HazelcastOperandTypeInference.java [hunk 3]
```diff
@@ -61,11 +64,13 @@ public void inferOperandTypes(SqlCallBinding callBinding, RelDataType returnType
...
                 if (parameter != null) {
                     SqlTypeName parameterType = parameter.type();
-                    operandTypes[i] = toType(parameterType, typeFactory);
+                    parameterTypes[parameter.ordinal()] = toType(parameterType, typeFactory);
                 } else {
                     throw SqlUtil.newContextException(id.getParserPosition(), RESOURCE.unknownArgumentName(name));
                 }
             }
+            //noinspection ResultOfMethodCallIgnored
+            Arrays.stream(parameterTypes).filter(Objects::nonNull).toArray(ignored -> operandTypes);
```
Comment (ivanthescientist) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/operators/typeinference/HazelcastOperandTypeInference.java:73]: What does this expression do actually?

**Review by ivanthescientist — APPROVED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/operators/typeinference/HazelcastOperandTypeInference.java (hunk 3 above)
Comment (gierlachg) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/operators/typeinference/HazelcastOperandTypeInference.java:73]: Copies over non-null elements. `callBinding` & `operandTypes` respectively contains and has capacity for actually used parameters.

Comment (viliam-durina) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/operators/typeinference/HazelcastOperandTypeInference.java:73]: I'd also prefer to see this as a for loop, this is hard to read...

**Pull request merged**